polyply/src/annotate_dna.py

```diff
@@ -0,0 +1,194 @@
+        includes_DNA: list[int]
+            node keys list of the meta_molecules tagged as indluding DNA
+        """
+        self.topology = topology
+        self.includes_DNA = [float(ndx) for ndx in includes_DNA]
+
+    def _find_dna_strands(self, meta_molecule):
+        """
+        Extract the subgraphs of meta_molecule that are DNA strands
+
+        Parameters
+        ------------
+        meta_molecule: :class:`polyply.src.MetaMolecule`
+
+        Returns
+        ---------
+        strands: list[:class: `networkx.Graph`]
+            list of DNA strands represented as subgraphs of meta_molecule
+        """
+
+        def _is_nucleobase(node):
+            return (meta_molecule.nodes[node]['resname'] in base_library)
+
+        seperated_meta_mol = [meta_molecule.subgraph(graph) for graph
+                              in nx.connected_components(meta_molecule)]
+
+        # Find all strands by performing greedy search on
+        # all connected_components of meta_molecule
+        strands = []
+        for meta_mol in seperated_meta_mol:
+            queue = sample(meta_mol.nodes, 1)
+            visited = []
+            strand = []
+            while queue:
+                # pop first node out of queue
+                current_node = queue.pop()
+                visited.append(current_node)
+
+                if _is_nucleobase(current_node):
+                    strand.append(current_node)
+                elif strand:
+                    strands.append(meta_mol.subgraph(strand))
+                    strand = []
+
+                # Determine neighbors nodes
+                neighbors = [node for node in meta_molecule.neighbors(current_node) if node not in queue + visited]
+                # update queue
+                queue = neighbors + queue
+                # Sort queue according to heuristic
+                queue.sort(key=_is_nucleobase, reverse=True)
+
+            if strand not in strands:
+                strands.append(meta_molecule.subgraph(strand))
+        return strands
+
+    def _determine_sequence(self, strand):
+        sequence = []
+        if nx.cycle_basis(strand):
+            start_base = sample(strand.nodes, 1)
+        else:
+            condition = lambda node: strand.degree(node) == 1
+            start_base = next(filter(condition, strand.nodes), None)
+        for resid in nx.dfs_tree(strand, start_base):
+            sequence.append(strand.nodes[resid]['resname'])
+        return sequence
+
+    def _find_complementary_strands(self, ref_strand, strands):
+        # Sequence we want to find match for
+        ref_seq = self._determine_sequence(ref_strand)
+
+        match_ratios = {}
+        for ndx, strand in enumerate(strands, start=1):
+            seq = self._determine_sequence(strand)
+
+            match_ratio = 0
+            for ref_base, base in zip(ref_seq, seq):
+                if base_library[ref_base] == base:
+                    match_ratio += 1
```

Comment:
this should be it's own function taking ref_seq and seq and returning a match ratio